README.md
# Whack-a-Devil

demo: https://youtu.be/NAAmeuHo6Sc

**Dependencies**:
- better-sqlite3
- browser-sync
- bulma
- concurrently
- cors
- express
- md5
- requirejs

**How to Run**: npm run start

**Find details in [docs](/docs) on**:
- back end user databases
- changes
- planning

## The Game
The game is a modified version of "Whack a Mole"

**Sign up**

Create your profile in the sign up section if you don't already have an account

**Login**

Login with registered username and password

**Gameplay**

Once loggged in, you will be directed to the game itself. 
The game is a modified version of "Whack a Mole" with a 60 sec countdown.
You receive one score by clicking on the devil img.
Your goal is to score as high as possible

**Features**
- Show High Score: when you click this button, the highest score shows up with the name of the player
- Save Score: when clicked, saves your current score
- Change Username: when clicked, changes your username and logs you out
- Delete Account: when clicked, deletes your account permanently
- View Profile: view profile details, including: User, User Name, Email, and Year

## HOW IT WORKS

#### Front End
The front end development and styling of the interface was done using Bulma. We added divs and classes to index.html so that we could create separation in the "Create An Account", "Change Username", and "Delete Account" parts using boxes. We also added styling through using different fonts, colors, and spacing to make our interface more appealing to the user. The Whack-A-Devil gameboard was also altered using Bulma to have Carolina Blue background instead of it's original formatting in black and white.

#### Gameplay
You are given 60 seconds to hit the Duke devil as many times as you can. At the end of the 60 seconds, you score is outputted

#### Back end
Both the back-end databases were created using Sqlite-3, md5, and express, as done in assignments 4 and 5. CRUD API endpoints were created for each of the user and score functions defined above. 

##### User database:
Using input forms and buttons, users made HTTP requests to the database to access, update, and delete their info. The database, user.db, has 5 parameters: username, password, email, name, and year in school. Of these, username, password, and email are required for each new account. Additionally, changes had to be made in the database to require that each username was unique, so there could be no duplicate users. The API endpoints had functions to create a new user, log in, view user info, change username, delete account, and log out of account. The back end user database files also had scripts to manage which sections of the website were shown based on the log in status. Further documentation of this database and its associated functions can be found within the src folder's files of database.js, forms.js, and server.js, or in the back-end-userdatabase.md file in the docs folder.

##### Score database:
The scoreboard database is created along with the user database. The score database consists of two fields: Name and Score. 
When the "Get Highest Score" button is clicked, the "getHighest()" function in forms.js sends a request to the server, accessing the http://localhost:5000/app/user/highest endpoint, pulling the highest score and the name of the player from the scoreboard database.
When the "Save Score" button is clicked, the "saveScore()" function in forms.js sends a request to the server, accessing the http://localhost:5000/app/new/score endpoint, saving your current score and your name into the scoreboard database


docs/changes.md
# Future Potential Changes

### Game Changes
- can click devil as many times as you want, needs to be delay
- after final score, everything should stop
- start/end button

### Front End Changes
- make timer countdown and game on same page
- make an animated devil slide up instead of flat 3x3 grid

### Back End Changes
- more form validation
- show multiple high scores

src/app.js
const square = document.querySelectorAll('.square')
const blue_devil = document.querySelectorAll('.blue_devil')
const timeLeft = document.querySelector('#seconds-left')
let score = document.querySelector('#score')
let highest_score = document.querySelector('#highest_score')
let player = document.querySelector('#player')


// Connect to a database or create one if it doesn't exist yet.

let result = 0
let currentTime = timeLeft.textContent

function randomSquare() {
    square.forEach(className => {
        className.classList.remove('blue_devil')
    })
    
    let randomPosition = square[Math.floor(Math.random() * 9)]
    randomPosition.classList.add('blue_devil')

    //assin the id of the randomPosition to hitPosition for us to use later
    hitPosition = randomPosition.id
}

square.forEach(id => {
    id.addEventListener('mouseup', () => {
        if(id.id === hitPosition) {
            result++
            score.textContent = result
        }
    })
})

function moveDevil() {
    let timerId = null
    timerId = setInterval(randomSquare, 1000)
}
moveDevil()

function countDown() {
    currentTime--
    timeLeft.textContent = currentTime

    if (currentTime == 0) {
        clearInterval(timerId);
        alert('GAME OVER! Your final score is ' + result);
    }
}

let timerId = setInterval(countDown, 1000)

src/server.js
// Define app using express
var express = require("express");
var app = express();

// Require database SCRIPT files
var db = require("./database.js");

var scoreboard = require("./score_database.js");

// Require md5 MODULE
var md5 = require("md5");

// Require CORS module
var cors = require("cors");

// Make Express use its own built-in body parser
app.use(express.urlencoded({ extended: true }));
app.use(express.json());

// Make Express use CORS
app.use(cors());

// Set server port
var HTTP_PORT = 5000;

// Start server
app.listen(HTTP_PORT, () => {
    console.log("Server running on port %PORT%".replace("%PORT%",HTTP_PORT))
});
// READ (HTTP method GET) at root endpoint /app/
app.get("/app/", (req, res, next) => {
    res.json({"message":"Your API works! (200)"});
	res.status(200);
});

// Define other CRUD API endpoints using express.js and better-sqlite3
// CREATE a new user (HTTP method POST) at endpoint /app/new/
app.post("/app/new/user", (req, res, next) => {
	var data = {
		user: req.body.user,
		email: req.body.email,
		pass: req.body.pass ? md5(req.body.pass) : null,
		name: req.body.name, 
		year: req.body.year
	}
	const stmt = db.prepare('INSERT INTO userinfo (user, pass, email, name, year) VALUES (?, ?, ?, ?, ?)');
	const info = stmt.run(data.user, data.pass, data.email, data.name, data.year);
	res.status(201).json({"message":info.changes+" record created: ID "+info.lastInsertRowid+" (201)"});
})

// CREATE a new score (HTTP method POST) at endpoint /app/new/score               ******not working
app.post("/app/new/score", (req, res) => {
	var data = {
		user: req.body.user,
		name: req.body.name, 
		score: req.body.score
	}
	const stmt = scoreboard.prepare('INSERT INTO scoreboard (user, name, score) VALUES (?, ?, ?)');
	const info = stmt.run(data.user, data.name, data.score);
	res.status(201).json({"message":info.changes+" score record created: ID "+info.lastInsertRowid+" (201)"});
})

// READ a list of all users (HTTP method GET) at endpoint /app/users/
app.get("/app/users", (req, res) => {	
	const stmt = db.prepare("SELECT * FROM userinfo").all();
	res.status(200).json(stmt);
});


// READ a list of all score data (HTTP method GET) at endpoint /app/scores/  **** not working
app.get("/app/scores", (req, res) => {	
	const stmt = scoreboard.prepare("SELECT * FROM scoreboard").all();
	res.status(200).json(stmt);
});


// READ a list of one user that has the highest score (HTTP method GET) at endpoint /app/user/highest
app.get("/app/user/highest", (req, res) => {	
	const stmt = scoreboard.prepare("SELECT name, score FROM scoreboard ORDER BY score DESC LIMIT 5;").get();
	res.status(200).json(stmt);
});

//SELECT id, num1 + num2 AS total FROM table1 ORDER BY num1 + num2 DESC LIMIT 5

// UPDATE a single user (HTTP method PATCH) at endpoint /app/updating/user/
app.patch("/app/updating/user", (req, res) => {
	var data = {
		user: req.body.user,
		pass: req.body.pass ? md5(req.body.pass) : null,
		newuser: req.body.newuser
	}
	const stmt = db.prepare("UPDATE userinfo SET user = COALESCE(?,user) WHERE user = ? AND pass = ?")
	const info = stmt.run(data.newuser, data.user, data.pass)
	res.status(200).json({"message":`1 record updated: User ${data.user} (200)`});
});


// DELETE a single user by username and password (HTTP method DELETE) at endpoint /app/deleting/user
app.delete("/app/deleting/user", (req, res) => {
    var data = {
		user: req.body.user,
		pass: req.body.pass ? md5(req.body.pass) : null
	}
	const stmt = db.prepare("DELETE FROM userinfo WHERE user = ? AND pass = ?")
    const info = stmt.run(data.user, data.pass);
	res.status(200).json({"message":`1 record deleted: User ${data.user} (200)`});
});


// READ a single user (HTTP method GET) at endpoint /app/login/user
app.post("/app/login/user", (req, res) => {	
    var data = {
		user: req.body.user,
		pass: req.body.pass ? md5(req.body.pass) : null
	}
	const stmt = db.prepare("SELECT * FROM userinfo WHERE user = ? AND pass = ?").get(data.user, data.pass);
	res.status(200).json(stmt);
});


// READ a single user (HTTP method GET) at endpoint /app/exists/user
app.post("/app/exists/user", (req, res) => {	
    var data = {
		user: req.body.user
	}
	const stmt = db.prepare("SELECT * FROM userinfo WHERE user = ?").get(data.user);
	res.status(200).json(stmt);
});

// Default response for any other request
app.use(function(req, res){
	res.json({"message":"Your API is working!"});
    res.status(404);
});


src/database.js
// This ensures that things do not fail silently but will throw errors instead.
"use strict";
// Require better-sqlite.
const Database = require('better-sqlite3');

// Connect to a database or create one if it doesn't exist yet.
var db = new Database('src/user.db');

// Is the database initialized or do we need to initialize it?
const stmt = db.prepare(`SELECT name FROM sqlite_master WHERE type='table' and name='userinfo';`);
let row = stmt.get();
if (row === undefined) {
// Echo information about what you are doing to the console.
    console.log('Your user database appears to be empty. Initializing it now.');
// Set a const that will contain your SQL commands to initialize the database.
// Include constraint that user must be unique
    const sqlInit = `
        CREATE TABLE userinfo ( id INTEGER PRIMARY KEY, user TEXT, pass TEXT, email TEXT, name TEXT, year TEXT, unique(user) );
		INSERT INTO userinfo (user, pass, email, name, year) VALUES ('admin','bdc87b9c894da5168059e00ebffb9077', 'admin@test.com', 'Sam Anthony', '2023');
    `;
// Execute SQL commands that we just wrote above.
    db.exec(sqlInit);
// Echo information about what we just did to the console.
    console.log('Your database has been initialized with a new table and three entries containing a username, email address, and password');
} else {
// Since the database already exists, echo that to the console.
    console.log('Database exists.')
}
// Export all of the above as a module so that we can use it elsewhere.
module.exports = db

src/score_database.js
// This ensures that things do not fail silently but will throw errors instead.
"use strict";
// Require better-sqlite.
const Database = require('better-sqlite3');

// Connect to a database or create one if it doesn't exist yet.
var scoreboard = new Database('src/scoreboard.db');

// Is the database initialized or do we need to initialize it?
const stmt = scoreboard.prepare(`SELECT name FROM sqlite_master WHERE type='table' and name='scoreboard';`);
let row = stmt.get();
if (row === undefined) {
// Echo information about what you are doing to the console.
    console.log('Your scoreboard database appears to be empty. Initializing it now.');
// Set a const that will contain your SQL commands to initialize the database.
// Include constraint that user must be unique
    const sqlInit = `
        CREATE TABLE scoreboard ( id INTEGER PRIMARY KEY, user TEXT, name TEXT, score INTEGER);
        INSERT INTO scoreboard (user, name, score) VALUES ('test3','Test Three', 10);
        INSERT INTO scoreboard (user, name, score) VALUES ('admin','Sam Anthony', 0);
        INSERT INTO scoreboard (user, name, score) VALUES ('test2','Test TWO', 5);
        INSERT INTO scoreboard (user, name, score) VALUES ('test4','Test FOUR', 15);
        INSERT INTO scoreboard (user, name, score) VALUES ('test5','Test FIVE', 7);
        INSERT INTO scoreboard (user, name, score) VALUES ('test6','Test SIX', 20);
    `;
// Execute SQL commands that we just wrote above.
    scoreboard.exec(sqlInit);
// Echo information about what we just did to the console.
    console.log('Your database has been initialized with a new table and three entries containing a username, name, and initial high score of 0.');
} else {
// Since the database already exists, echo that to the console.
    console.log('Database exists.')
}
// Export all of the above as a module so that we can use it elsewhere.
module.exports = scoreboard

src/forms.js
window.addEventListener("load", function () {
    //  Global User Vars
    var thisUser = null;
    var loggedIn;

    // Hide and show elements based on login status
    function hideElements(){
        document.getElementById("loggedInUser").classList.add("hide-on-logout");
        document.getElementById("scorebox").classList.add("hide-on-logout");
        document.getElementById("game").classList.add("hide-on-logout");
        document.getElementById("signupdiv").classList.remove("hide-on-logout");
        document.getElementById("loginmessage").classList.remove("hide-on-logout");
        document.getElementById("credentials").classList.remove("hide-on-logout");
    }
    function showElements(){
        document.getElementById("loggedInUser").classList.remove("hide-on-logout");
        document.getElementById("scorebox").classList.remove("hide-on-logout");
        document.getElementById("game").classList.remove("hide-on-logout");
        document.getElementById("signupdiv").classList.add("hide-on-logout");
        document.getElementById("loginmessage").classList.add("hide-on-logout");
        document.getElementById("credentials").classList.add("hide-on-logout");
    }

    // Remove this line to show elements at start before login during testing
    hideElements();
    

    // New User Creation
    function sendData( form ) {
        const sendRequest = new XMLHttpRequest();
        const signupInfo = new URLSearchParams(new FormData( form ));
        sendRequest.addEventListener("error", function(event){
            alert('Submission unsuccessful! Please try again.');
        });
        sendRequest.addEventListener("load", function(event){
            alert('Your account was created!');
        });
        sendRequest.open("POST", "http://localhost:5000/app/new/user");
        sendRequest.send( signupInfo );
    }

    function newUserValid() {
        if(document.getElementById("signup").user.value == "" | document.getElementById("signup").email.value == "" | document.getElementById("signup").pass.value == ""){
            return false;
        } else {
            return true;
        }
    }

    const newuser = document.getElementById("signup");
    newuser.addEventListener("submit", function(event){
        event.preventDefault();
        if(newUserValid()){
            sendData(this);
        } else {
            alert("Username, Email, and Password are Required!")
        }
        
    });

    // Delete User
    function deleteData( form ) {
        var deletePassword = form.pass.value;
        let deleteInfo = new URLSearchParams("");
        deleteInfo.append('user', thisUser.user);
        deleteInfo.append('pass', deletePassword);
        const sendRequest = new XMLHttpRequest();
        sendRequest.addEventListener("error", function(event){
            alert('Deletion unsuccessful! Please try again.');
        });
        sendRequest.addEventListener("load", function(event){
            // alert('Your account was deleted!');
        });

        sendRequest.open("DELETE", "http://localhost:5000/app/deleting/user");
        sendRequest.send( deleteInfo );
    }

    const olduser = document.getElementById("delete");
    olduser.addEventListener("submit", function(event){
        event.preventDefault();
        deleteData(this);
    });

    // Updating User
    function updateData( form ) {
        const sendRequest = new XMLHttpRequest();
        var updateInfo = new URLSearchParams(new FormData( form ));
        updateInfo.append('user', thisUser.user);
        sendRequest.addEventListener("error", function(event){
            alert('Changes were unsuccessful! Please try again.');
        });
        sendRequest.addEventListener("load", function(event){
            // alert('Your username was changed!');
        });
        sendRequest.open("PATCH", "http://localhost:5000/app/updating/user");
        sendRequest.send( updateInfo );
    }

    const updateUser = document.getElementById("changeName");
    updateUser.addEventListener("submit", function(event){
        event.preventDefault();
        updateData(this);
    });


    // Get List of Users
    // function getUsers( form ) {
    //     const sendRequest = new XMLHttpRequest();
    //     sendRequest.addEventListener("error", function(event){
    //         alert('Accessing users unsuccessful! Please try again.');
    //     });

    //     sendRequest.onreadystatechange = function() {
    //         if (this.readyState == 4 && this.status == 200) {
    //             // alert(sendRequest.responseText);
    //             document.getElementById("allUsers").innerHTML = sendRequest.responseText;
    //         }
    //     }
    //     sendRequest.open("GET", "http://localhost:5000/app/users");
    //     sendRequest.send();        
    // };

    // const allUsers = document.getElementById("usersButton");
    // allUsers.addEventListener("click", function(event){
    //     event.preventDefault();
    //     getUsers(this)
    // });


    // Login
    function getUserdata( form ) {
        const sendRequest = new XMLHttpRequest();
        const userInfo = new URLSearchParams(new FormData( form ));
        sendRequest.addEventListener("error", function(event){
            alert('Accessing users unsuccessful! Please try again.');
        });

        sendRequest.onreadystatechange = function() {
            if (this.readyState == 4 && this.status == 200) {
                thisUser = JSON.parse(sendRequest.response);
                if(loggedIn){
                    thisUser = null;
                    loggedIn = false;
                    document.getElementById("login").value = "Login";
                    document.getElementById("greeting").innerHTML = thisUser;
                    // document.getElementById("loggedInUser").classList.add("hide-on-logout");
                    hideElements()
                    alert("Logging Out");
                } else {
                    if(thisUser.name == null | thisUser.name == ""){
                        document.getElementById("greeting").innerHTML = "Welcome to Whack-a-Devil, " + thisUser.user;
                    } else {
                        document.getElementById("greeting").innerHTML = "Welcome to Whack-a-Devil, " + thisUser.name;
                    };
                    document.getElementById("login").value = "Logout";
                    loggedIn = true;
                    // document.getElementById("loggedInUser").style.display = "block";
                    // document.getElementById("loggedInUser").classList.remove("hide-on-logout");
                    showElements();
                    alert("Login successful!")
                }
            }
        }
        sendRequest.open("POST", "http://localhost:5000/app/login/user");
        sendRequest.send( userInfo );
    };

    const selfUser = document.getElementById("loginForm");
    selfUser.addEventListener("submit", function(event){
        event.preventDefault();
        if(loggedIn){
            alert("Logging Out");
            // document.getElementById("loggedInUser").classList.add("hide-on-logout");
            hideElements()
            thisUser = null;
            loggedIn = false;
            document.getElementById("login").value = "Login";
            document.getElementById("greeting").innerHTML = thisUser;
            this.reset();
        } else {
            getUserdata(this);
        }
    });

    // Show Profile
    const showData = document.getElementById("showData");
    showData.addEventListener("click", function(event){
        event.preventDefault();
        if(loggedIn){
            document.getElementById("profileData").innerHTML = `Username: ${thisUser.user}, 
            Email: ${thisUser.email}, 
            Name: ${thisUser.name},
            Year: ${thisUser.year}`
        } else {
            alert("You must log in to see profile!")
        }
    });

    // create a score profile after each time  ******not working
    function saveScore( form, score ) {
        const sendRequest = new XMLHttpRequest();
        var updateInfo = new URLSearchParams(new FormData( form ));
        updateInfo.append('user', thisUser.user);
        updateInfo.append('name', thisUser.name);
        updateInfo.append('score', score);
        sendRequest.addEventListener("error", function(event){
            alert('Changes were unsuccessful! Please try again.');
        });
        sendRequest.addEventListener("load", function(event){
            // alert('Your score was added!');
        });
        sendRequest.open("POST", "http://localhost:5000/app/new/score");
        sendRequest.send( updateInfo );
    }

    // when click "save your score" button, do saveScore()
    const score_val = document.getElementById("score");
    const score_button = document.getElementById("saveScore");
    score_button.addEventListener("click", function(event){
        event.preventDefault();
        if (loggedIn){
            saveScore(this, score_val.innerHTML); 
        } else {
            alert('Log in to save score!');
        }
    });


    // retrieve the highest score from database
    function getHighest( form ) {
        const sendRequest = new XMLHttpRequest();
        sendRequest.addEventListener("error", function(event){
            alert('retrieving score unsuccessful! Please try again.');
        });

        sendRequest.onreadystatechange = function() {
            if (this.readyState == 4 && this.status == 200) {
                // alert(sendRequest.responseText);
                document.getElementById("showScore").innerHTML = sendRequest.responseText;
            }
        }
        sendRequest.open("GET", "http://localhost:5000/app/user/highest");
        sendRequest.send(); 
    }
// when click "show highest score" button, do getHighest()
    const getScore = document.getElementById("highestScore");
    getScore.addEventListener("click", function(event){
        event.preventDefault();
        getHighest(this)        
    });

});    
